File: mycollab-web/src/main/java/com/esofthead/mycollab/vaadin/ui/TooltipBuilder.java
/**
 * This file is part of mycollab-web.
 *
 * mycollab-web is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-web is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-web.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.esofthead.mycollab.vaadin.ui;

import com.esofthead.mycollab.core.utils.StringUtils;
import com.hp.gagawa.java.elements.A;
import com.hp.gagawa.java.elements.Div;
import com.hp.gagawa.java.elements.H3;
import com.hp.gagawa.java.elements.Img;
import com.hp.gagawa.java.elements.Table;
import com.hp.gagawa.java.elements.Td;
import com.hp.gagawa.java.elements.Tr;

/**
 * 
 * @author MyCollab Ltd.
 * @since 4.1.1
 * 
 */
public class TooltipBuilder {

	private Div div;
	private Table table;

	public TooltipBuilder() {
		div = new Div();
		div.setStyle("font: 12px Arial, Verdana, Helvetica, sans-serif !important;line-height: normal;");
		table = new Table();
		table.setStyle("padding-left:10px; width :500px; color: #5a5a5a; font-size:11px;");
		div.appendChild(table);
	}

	public TooltipBuilder setTitle(String title) {
		H3 htmlTitle = new H3();
		htmlTitle.appendText(title);
		div.appendChild(0, htmlTitle);
		return this;
	}

	public TooltipBuilder appendRow(Tr tr) {
		table.appendChild(tr);
		return this;
	}

	public Div create() {
		return div;
	}

	public static class TdUtil {
		public static Td buildCellName(String name) {
			return new Td().setStyle(
					"width: 100px; vertical-align: top; text-align: right;")
					.appendText(name + ": ");
		}

		public static Td buildCellValue(String value) {
			String cutNameVal = StringUtils.getStringFieldValue(value);
			return new Td()
					.setStyle(
							"width:200px;word-wrap: break-word; white-space: normal;vertical-align: top; word-break: break-all;")
					.appendText(cutNameVal);
		}

		public static Td buildCellValue(Number value) {
			return (value == null) ? buildCellValue("") : buildCellValue(value
					.toString());

		}

		public static Td buildCellLink(String href, String name) {
			String cutNameVal = StringUtils.getStringFieldValue(name);
			return new Td()
					.setStyle(
							"width:200px;word-wrap: break-word; white-space: normal;vertical-align: top; word-break: break-all;")
					.appendChild(new A().setHref(href).appendText(cutNameVal));
		}

		public static Td buildCellLink(String href, String imageLink,
				String name) {
			String cutNameVal = StringUtils.getStringFieldValue(name);
			return new Td()
					.setStyle(
							"width:200px;word-wrap: break-word; white-space: normal;vertical-align: top; word-break: break-all;")
					.appendChild(
							new A().setHref(href)
									.appendChild(new Img("", imageLink))
									.appendText(cutNameVal));
		}
	}
}

File: mycollab-core/src/main/java/com/mycollab/html/TooltipBuilder.kt
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.html

import com.hp.gagawa.java.elements.*
import com.mycollab.core.utils.StringUtils

/**
 * @author MyCollab Ltd
 * @since 6.0.0
 */
class TooltipBuilder() {
    private var table = Table().setStyle("padding-left:10px; width:550px; color: black; font-size:12px;")
    private var div = Div().setStyle("line-height: normal;").appendChild(table)

    fun appendTitle(title: String): TooltipBuilder {
        val htmlTitle = H3().setStyle("whitespace: normal; width: 100%; word-wrap: break-word;").appendText(title)
        div.appendChild(0, htmlTitle)
        return this;
    }

    fun appendRow(tr: Tr): TooltipBuilder {
        table.appendChild(tr)
        return this
    }

    fun create(): Div = div

    companion object {
        @JvmStatic fun buildCellName(name: String): Td = Td().setStyle("width: 120px; vertical-align: top; text-align: right; color:#999").
                appendText(name + ": ")

        @JvmStatic fun buildCellValue(value: String?): Td {
            val cutNameVal = StringUtils.trimHtmlTags(value)
            return Td().setStyle("width:200px;word-wrap: break-word; white-space: normal;vertical-align: top;").appendText(cutNameVal)
        }

        @JvmStatic fun buildCellValue(value: Number?): Td = if (value == null) buildCellValue("") else buildCellValue(value.toString())

        @JvmStatic fun buildCellLink(href: String?, name: String?): Td {
            val cutNameVal = StringUtils.trimHtmlTags(name)
            return Td().setStyle("width:200px;word-wrap: break-word; white-space: normal;vertical-align: top;").
                    appendChild(A().setHref(href).appendText(cutNameVal))
        }

        @JvmStatic fun buildCellLink(href: String, imageLink: String, name: String?): Td {
            val cutNameVal = StringUtils.trimHtmlTags(name)
            return Td().setStyle("width:200px;word-wrap: break-word; white-space: normal;vertical-align: top;").
                    appendChild(Img("", imageLink)).appendChild(DivLessFormatter.EMPTY_SPACE).
                    appendChild(A().setHref(href).appendText(cutNameVal))
        }
    }
}